firstpartydata/first_party_data.go

```diff
@@ -0,0 +1,354 @@
+	newSite.Page = fpdConfigSite.Page
+	newSite.Search = fpdConfigSite.Search
+	newSite.Keywords = fpdConfigSite.Keywords
+
+	if len(fpdConfigSite.Ext) > 0 {
+		if len(originalSite.Ext) > 0 {
+			newSite.Ext, err = jsonpatch.MergePatch(originalSite.Ext, fpdConfigSite.Ext)
+		} else {
+			newSite.Ext = fpdConfigSite.Ext
+		}
+	}
+
+	return newSite, err
+}
+
+func resolveApp(fpdConfigApp *openrtb2.App, bidRequestApp *openrtb2.App, globalFPD map[string][]byte, openRtbGlobalFPD map[string][]openrtb2.Data, hasGlobalFPD bool) (*openrtb2.App, error) {
+
+	if bidRequestApp == nil && fpdConfigApp == nil {
+		return nil, nil
+	}
+	if bidRequestApp == nil {
+		bidRequestApp = &openrtb2.App{}
+	}
+	newApp := *bidRequestApp
+	var err error
+
+	if hasGlobalFPD {
+		//apply global fpd if exists
+		if len(globalFPD[appKey]) > 0 {
+			extData := buildExtData(globalFPD[appKey])
+			if len(newApp.Ext) > 0 {
+				newApp.Ext, err = jsonpatch.MergePatch(newApp.Ext, extData)
+			} else {
+				newApp.Ext = extData
+			}
+		}
+		if len(openRtbGlobalFPD[appContentDataKey]) > 0 {
+			if newApp.Content != nil {
+				contentCopy := *newApp.Content
+				contentCopy.Data = openRtbGlobalFPD[appContentDataKey]
+				newApp.Content = &contentCopy
+			} else {
+				newAppContent := &openrtb2.Content{Data: openRtbGlobalFPD[appContentDataKey]}
+				newApp.Content = newAppContent
+			}
+		}
+	}
+
+	if fpdConfigApp != nil {
+		//apply bidder specific fpd if present
+		newApp, err = mergeApps(&newApp, fpdConfigApp)
+	}
+
+	return &newApp, err
+}
+
+func mergeApps(originalApp *openrtb2.App, fpdConfigApp *openrtb2.App) (openrtb2.App, error) {
+
+	var err error
+	newApp := openrtb2.App{}
+	newApp = *originalApp
+
+	newApp.Name = fpdConfigApp.Name
+	newApp.Bundle = fpdConfigApp.Bundle
+	newApp.Domain = fpdConfigApp.Domain
+	newApp.StoreURL = fpdConfigApp.StoreURL
+	newApp.Cat = fpdConfigApp.Cat
+	newApp.SectionCat = fpdConfigApp.SectionCat
+	newApp.PageCat = fpdConfigApp.PageCat
+	newApp.Ver = fpdConfigApp.Ver
+	newApp.Keywords = fpdConfigApp.Keywords
+
+	if len(fpdConfigApp.Ext) > 0 {
+		if len(originalApp.Ext) > 0 {
+			newApp.Ext, err = jsonpatch.MergePatch(originalApp.Ext, fpdConfigApp.Ext)
+		} else {
+			newApp.Ext = fpdConfigApp.Ext
+		}
+	}
+
+	return newApp, err
+}
+
+func buildExtData(data []byte) []byte {
+	res := []byte(`{"data":`)
+	res = append(res, data...)
+	res = append(res, []byte(`}`)...)
+	return res
+}
+
+func PreprocessBidderFPD(reqExtPrebid openrtb_ext.ExtRequestPrebid) (map[openrtb_ext.BidderName]*openrtb_ext.FPDData, openrtb_ext.ExtRequestPrebid) {
+	//map to store bidder configs to process
+	fpdData := make(map[openrtb_ext.BidderName]*openrtb_ext.FPDData)
+
+	//every bidder in ext.prebid.data.bidders should receive fpd data if defined
+	bidderTable := make(map[string]struct{}) //just need to check existence of the element in map
+
+	if reqExtPrebid.Data != nil && len(reqExtPrebid.Data.Bidders) != 0 {
+
+		for _, bidder := range reqExtPrebid.Data.Bidders {
```

Comment:
Move this logic to the resolution function

Reply:
Refactored. 
Fun fact: it worked already due to previous refactoring when I removed "kill switch" feature (all bidders in Data.Bidders should be present in BidderConfigs). This functionality didn't break the logic, but definitely was redundant. Good catch!